client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -203,8 +214,41 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
-        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        while (true) {
+            synchronizer.fetchUpdates();
+            val state = synchronizer.getState();
+            // We only move into the block if the state transition has happened successfully.
+            if (stateTransition(state, new UpdatingConfig(true))) {
+                // Use the latest generation and reader group Id.
+                ReaderGroupConfig newConfig = config.toBuilder()
+                        .readerGroupId(state.getConfig().getReaderGroupId())
+                        .generation(state.getConfig().getGeneration()).build();
```

Comment:
Doesn't the generation need to change somewhere?

Reply:
The generation is updated on the controller. We update the config's generation on to the controller using the periodic update that we have on the `ReaderGroupStateManager`.

Reply:
The generation that is written into the `StateSynchronizer` is the one returned after `controller.updateReaderGroup` has been called.